### ISSUE TITLE
Eliminate NonCopyable approach

### ISSUE THREAD

**Issue by even1024** [epam/Indigo#778]
NonCopyable class should be removed from Indigo.
All the derived classes should have proper deep copy constructors.
![non_copyable__inherit__graph](https://user-images.githubusercontent.com/7814977/179456952-b94bf718-df56-40ff-9367-770f47e9e3c9.png)

**Comment by AlexanderSavelyev** [comment]
non-copyable was created for a performance reason. The task should not be done without performance impact evaluation for the Indigo calculations. Have proper deep copy constructors - for all classes assumes that someone can have copy objects where it is not necessary. Alternative possible option is to have move constructors defined, but not copy. 
Deprioritizing the task until providing more details
